File: python/qianfan/resources/http_client.py
# Copyright (c) 2023 Baidu, Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Any, AsyncIterator, Dict, Iterator, Optional, Tuple

import aiohttp
import requests

from qianfan import get_config
from qianfan.resources.typing import QfRequest


class HTTPClient(object):
    """
    object used to make http request
    """

    def __init__(self, ssl: bool = True, **kwargs: Any) -> None:
        """
        init sync and async request session

        Args:
            ssl (bool):
                whether to use ssl verification in connection,
                default to True
            **kwargs (Any):
                arbitrary arguments
        """
        cfg = get_config()
        if not ssl or not cfg.SSL_VERIFICATION_ENABLED:
            self.ssl = False
        else:
            self.ssl = True
        self._session = requests.session()

    def _requests_proxy(self) -> Optional[Dict[str, str]]:
        """
        return proxy setting for requests
        """
        proxy_url = get_config().PROXY
        if len(proxy_url) == 0:
            return None
        return {"http": proxy_url, "https": proxy_url}

    def _aiohttp_proxy(self) -> Optional[str]:
        """
        return proxy setting for requests
        """
        proxy_url = get_config().PROXY
        if len(proxy_url) == 0:
            return None
        return proxy_url

    def request(self, req: QfRequest) -> requests.Response:
        """
        sync request
        """
        resp = self._session.request(
            **req.requests_args(),
            timeout=req.retry_config.timeout,
            verify=self.ssl,
            proxies=self._requests_proxy(),
        )
        return resp

    def request_stream(
        self, req: QfRequest
    ) -> Iterator[Tuple[bytes, requests.Response]]:
        """
        sync stream request
        """
        resp = self._session.request(
            **req.requests_args(),
            stream=True,
            timeout=req.retry_config.timeout,
            verify=self.ssl,
            proxies=self._requests_proxy(),
        )
        for line in resp.iter_lines():
            yield line, resp

    async def arequest(
        self, req: QfRequest
    ) -> Tuple[aiohttp.ClientResponse, aiohttp.ClientSession]:
        """
        async request
        """
        session = aiohttp.ClientSession()
        timeout = aiohttp.ClientTimeout(total=req.retry_config.timeout)
        response = await session.request(
            **req.requests_args(),
            timeout=timeout,
            ssl=self.ssl,
            proxy=self._aiohttp_proxy(),
        )
        return response, session

    async def arequest_stream(
        self, req: QfRequest
    ) -> AsyncIterator[Tuple[bytes, aiohttp.ClientResponse]]:
        """
        async stream request
        """
        async with aiohttp.ClientSession() as session:
            timeout = aiohttp.ClientTimeout(total=req.retry_config.timeout)
            async with session.request(
                **req.requests_args(),
                timeout=timeout,
                ssl=self.ssl,
                proxy=self._aiohttp_proxy(),
            ) as resp:
                async for line in resp.content:
                    yield line, resp

File: java/src/main/java/com/baidubce/qianfan/util/http/HttpClient.java
/*
 * Copyright (c) 2024 Baidu, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.baidubce.qianfan.util.http;

import com.baidubce.qianfan.util.Json;
import com.baidubce.qianfan.util.TypeRef;
import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse;
import org.apache.hc.client5.http.impl.classic.HttpClients;
import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager;
import org.apache.hc.core5.http.*;
import org.apache.hc.core5.http.io.entity.EntityUtils;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;

public class HttpClient {
    private static final int MAX_CONNECTIONS = 128;

    private static final CloseableHttpClient client;

    static {
        PoolingHttpClientConnectionManager cm = new PoolingHttpClientConnectionManager();
        cm.setMaxTotal(MAX_CONNECTIONS);
        cm.setDefaultMaxPerRoute(MAX_CONNECTIONS);

        client = HttpClients.custom().setConnectionManager(cm).build();
    }

    private HttpClient() {
    }

    public static HttpRequest request() {
        return new HttpRequest();
    }

    public static <T> HttpResponse<T> executeJson(ClassicHttpRequest request, TypeRef<T> typeRef) throws IOException {
        return executeJson(request, typeRef.getType());
    }

    public static <T> HttpResponse<T> executeJson(ClassicHttpRequest request, Type type) throws IOException {
        return execute(request, (body, resp) -> {
            String stringBody = EntityUtils.toString(body);
            return resp
                    .setStringBody(stringBody)
                    .setBody(Json.deserialize(stringBody, type));
        });
    }

    public static HttpResponse<String> executeString(ClassicHttpRequest request) throws IOException {
        return execute(request, (body, resp) -> {
            String stringBody = EntityUtils.toString(body);
            return resp.setStringBody(stringBody)
                    .setBody(stringBody);
        });
    }

    public static HttpResponse<byte[]> execute(ClassicHttpRequest request) throws IOException {
        return execute(request, (body, resp) -> resp.setBody(EntityUtils.toByteArray(body)));
    }

    private static <T> HttpResponse<T> execute(ClassicHttpRequest request, HttpResponseBodyHandler<T> bodyHandler)
            throws IOException {
        return client.execute(request, resp -> {
            Map<String, String> headers = new LinkedHashMap<>();
            for (Header header : resp.getHeaders()) {
                headers.put(header.getName(), header.getValue());
            }
            HttpResponse<T> response = new HttpResponse<T>()
                    .setCode(resp.getCode())
                    .setHeaders(headers);
            return bodyHandler.handle(resp.getEntity(), response);
        });
    }

    public static HttpResponse<Iterator<String>> executeSSE(ClassicHttpRequest request) throws IOException {
        // Use legacy API to avoid auto-closing the response
        CloseableHttpResponse resp = client.execute(request);

        Map<String, String> headers = new LinkedHashMap<>();
        for (Header header : resp.getHeaders()) {
            headers.put(header.getName(), header.getValue());
        }

        Iterator<String> body = null;
        String stringBody = null;

        String contentType = headers.getOrDefault(ContentType.HEADER, "");
        if (contentType.startsWith(ContentType.TEXT_EVENT_STREAM)) {
            body = SSEWrapper.wrap(resp.getEntity().getContent(), resp);
        } else {
            try {
                // If the response is not an SSE stream, read the whole body as a string
                stringBody = EntityUtils.toString(resp.getEntity());
            } catch (ParseException e) {
                throw new IOException(e);
            }
        }

        return new HttpResponse<Iterator<String>>()
                .setCode(resp.getCode())
                .setHeaders(headers)
                .setBody(body)
                .setStringBody(stringBody);
    }

    @FunctionalInterface
    private interface HttpResponseBodyHandler<T> {
        HttpResponse<T> handle(HttpEntity entity, HttpResponse<T> response) throws HttpException, IOException;
    }
}
